### ISSUE TITLE
vm: transactions that were included into the block but can't be executed

### ISSUE THREAD

**Issue by dshulyak** [spacemeshos/go-spacemesh#3272]
Current vm behavior is to skip transaction in two cases:
- out of order nonce
- not enough funds to cover for max gas

There is an argument that we shouldn't skip them, but consume since they are taking space in the mesh. Pending on research

**Comment by countvonzero** [comment]
> - out of order nonce

this one should be obsolete after https://github.com/spacemeshos/go-spacemesh/pull/3282 is merged. as block will not be syntactically valid.

**Comment by dshulyak** [comment]
> this one should be obsolete after https://github.com/spacemeshos/go-spacemesh/pull/3282 is merged. as block will not be syntactically valid.

i meant - vm expects nonce to be 3, but transaction has 4. we shouldn't invalidate whole block, because it is very likely to happen after self-healing.

**Comment by dshulyak** [comment]
## vm expectations  
1. tx may not be parsable  
2. tx may not be verifiable  
3. tx may be parsable but fail verification  
4. tx may have lower out of order nonce   
5. tx may have higher out of order nonce
6. principal may have not funds to cover max_gas  
7. principal may not have funds to cover max_spend  
8. block may overflow gas limit budget  
9. tx may be included in the block multiple times  
9.1. in the same block  
9.2. in the next block  

## tx states
1-3 is a MALFORMED transaction. MALFORMED transaction cannot be consumed, and hence can't pay for being included into the block.  
4-9 is an INVALID transaction. INVALID transaction can be consumed and pay for gas. But probably not 4 and 9.  
transactions that were executed but failed are in FAILED state.  
otherwise transaction is in SUCCESS state.  

If protocol assumptions are satisfied there should be no MALFORMED or INVALID transactions in the block. But for all implementation considerations it makes no difference.

## skipping unnecessary validation

when tx is accepted into mempool it must be verified. it also allows to skip verification when tx is applied in the vm. this is also very important for performance, since verification costs is at least as much as simple tx execution.
however if tx type 3 may exist we need to optionally verify it again. this requires passing additional bit to indicate that tx is already verified or not.

so if current API is

```go
 Apply(lctx ApplyContext, txs []types.RawTx, blockRewards []types.AnyReward)
``` 
it will be changed to

```go
 Apply(lctx ApplyContext, txs []types.RawVerifiedTx, blockRewards []types.AnyReward)

type RawVerifiedTx struct {
  Raw RawTx
  Verified bool
}
```

## discussion
- how much gas is charged for INVALID transactions (4-8)? is it always max amount limited by max_gas?  
- Should type 4 be consumed?  
- should we try to execute INVALID txs 6-7 before charging gas? or gas should be charged without executing transaction if it can't cover max_gas + max_spend?  
- what should be done with tx that already has a receipt, regardless if it was consumed or not  
  - if we don't filter out such transaction then we will have multiple receipts for a single transaction. at most one of them will be successful. this feels like a mess  
  - is MALFORMED tx is a special case?  
- if Type 5 is consumed - account nonce is updated even if it was out of order?

**Comment by tal-m** [comment]
> i meant - vm expects nonce to be 3, but transaction has 4. we shouldn't invalidate whole block, because it is very likely to happen after self-healing.

In our nonce scheme, skipping is allowed. The invalid nonce case is if the vm expects nonce >= 4 but the tx nonce is 3. So case (5) above is actually valid.





> 4. tx may have lower out of order nonce
 
This is an INVALID transaction and can't pay gas (it could result from copying a previous valid transaction, for example).

> 6. principal may have not funds to cover max_gas
 
This is a VALID transaction. So the principal pays as much gas as it can, and the tx fails if it runs out of gas (it's also reasonable to fail without executing, but I don't see a good reason to do so at the moment).

> 7. principal may not have funds to cover max_spend
 
What is `max_spend`? 

> 8. block may overflow gas limit budget

Isn't this a syntactically-invalid block? Do we have a dynamic gas limit that depends on history?

 
> 9. tx may be included in the block multiple times
>   9.1. in the same block

This is a syntactically invalid block.

>   9.2. in the next block

This is an INVALID transaction treated the same way as a bad nonce (in fact, the nonce is *guaranteed* to be bad for the second copy --- that's what nonces are for).

**Comment by tal-m** [comment]
> what should be done with tx that already has a receipt, regardless if it was consumed or not

What does it mean for a tx to "have a receipt"?

**Comment by lrettig** [comment]
I attempted to layout the entire tx lifecycle here, including all cases. Hopefully this is helpful: https://docs.google.com/spreadsheets/d/1PNyKl4DuA17Ku85Ajl5A1lwkE9V7EXaYHJpkRxwyQ_U/edit

> when tx is accepted into mempool it must be verified. it also allows to skip verification when tx is applied in the vm. this is also very important for performance, since verification costs is at least as much as simple tx execution.
> however if tx type 3 may exist we need to optionally verify it again. this requires passing additional bit to indicate that tx is already verified or not.

Verification does need to be performed multiple times, since a malicious node could include an unverifiable tx in a proposal. At the very least it needs to be done upon adding the tx to the mempool and when a proposal is received. I don't understand why a tx that failed verification would ever be passed into the VM, though - why do we need this bit?

> * how much gas is charged for INVALID transactions (4-8)? is it always max amount limited by max_gas?

see spreadsheet linked above

> * Should type 4 be consumed?

what does "consumed" mean?

> * should we try to execute INVALID txs 6-7 before charging gas? or gas should be charged without executing transaction if it can't cover max_gas + max_spend?

I think we should just drop these and not charge any gas. We can revisit this decision in future.

> * what should be done with tx that already has a receipt, regardless if it was consumed or not

I also don't understand what this means. A tx cannot have a receipt unless it was included in a block.

>   * if we don't filter out such transaction then we will have multiple receipts for a single transaction. at most one of them will be successful. this feels like a mess

This should not happen

>   * is MALFORMED tx is a special case?

?

> * if Type 5 is consumed - account nonce is updated even if it was out of order?

See Tal's comments, above - we don't impose strict ordering on nonces, so I think the answer is yes

**Comment by dshulyak** [comment]
> What does it mean for a tx to "have a receipt"?

receipt is created after transaction was executed by vm.

> What is max_spend?

max_spend is an max amount of funds that can be transferred in the transaction.

---

> Verification does need to be performed multiple times, since a malicious node could include an unverifiable tx in a proposal. At the very least it needs to be done upon adding the tx to the mempool and when a proposal is received.

it doesn't explain why verification needs to be performed more than once.

>  I don't understand why a tx that failed verification would ever be passed into the VM, though - why do we need this bit?

what protocol should do if block that contains any MALFORMED transactions is considered as valid by the tortoise? do we vote against such block and not execute? in the last discussion there was an agreement that we vote for it and execute it. in this case unverifiable transaction should be expected by vm

 > what does "consumed" mean?

"consumed" is basically executed with consuming nonce and creating result(receipt). because in current vm implementation we are also skipping some transactions 

> I also don't understand what this means. A tx cannot have a receipt unless it was included in a block.

this is exactly about tx being included in the blocks multiple times. in the first block it was included and had MALFORMED status because it wasn't parsable. in a later block it was included again and was SUCCESSFULLY executed. if we allow to execute transaction that has MALFORMED status - then protocol will have 2 receipts for the same transaction.

**Comment by tal-m** [comment]
> it doesn't explain why verification needs to be performed more than once.

It's possible that an honest node dropped a transaction (because it failed verification), and then a malicious node proposed the same transaction. Since the honest node "forgets" dropped transactions, it would have to verify again.

> what protocol should do if block that contains any MALFORMED transactions is considered as valid by the tortoise? do we vote against such block and not execute? in the last discussion there was an agreement that we vote for it and execute it. in this case unverifiable transaction should be expected by vm

The critical question is what kind of "MALFORMED" you mean. If it is **syntactically**-invalid --- i.e., there's no possible state in which the transaction could be executed --- the block containing the transaction is also syntactically invalid, and we vote against it regardless of the Tortoise. 
In all other cases, we vote according to Tortoise rules. 

So it's possible that the VM will be executing a transaction that fails `verify` (since `verify` *does* depend on state --- for example, it can fail if the account has not been spawned yet, and succeed if it has). 

In this case, I think the rule should be simple:
* Any transaction that doesn't pay any gas (this includes those that fail `verify`) is considered **ineffective** if it's in a block.  It has no effect on state (including nonce state), so it's not "consumed" and should not generate a receipt.
* Any transaction that *does* pay gas (including those that fail due to insufficient gas, after passing `verify`) is consumed (and updates the nonce). 

A note regarding `max_spend`: `max_spend` is only an upper bound on spending, it's not an error to spend less (or to have less).

**Comment by lrettig** [comment]
> The critical question is what kind of "MALFORMED" you mean

Yes, let's be clear about terminology here. In the above-linked [spreadsheet](https://docs.google.com/spreadsheets/d/1PNyKl4DuA17Ku85Ajl5A1lwkE9V7EXaYHJpkRxwyQ_U/edit) I make a distinction between "malformed" (i.e., syntactically invalid) and "unverifiable" transactions (in addition to "duplicate" txs, which should also be dropped). Later in the lifecycle, we also have txs that "fail intrinsic gas check" and have "not enough funds" (both classes of which pay no gas).

> If it is **syntactically**-invalid --- i.e., there's no possible state in which the transaction could be executed --- the block containing the transaction is also syntactically invalid, and we vote against it regardless of the Tortoise.

I think we need to go even further than this. The node should totally refuse to execute such an invalid block even if the majority of weight votes for it in tortoise. The only way this could happen is either a "51% attack" or else after a contentious hard fork. Such a node is effectively on a minority fork that won't have any valid blocks for a while.

None of the txs in the block should be executed or marked as "known" or considered "consumed." There should _never_ be two receipts for the same txid. (The same tx could end up in two canonical blocks due to a reorg, or possibly due to another violation of security assumptions, but in this case, all txs after the first should be considered ineffective and be ignored completely, i.e., not consumed and no receipt.)

**Comment by tal-m** [comment]
> I think we need to go even further than this. The node should totally refuse to execute such an invalid block even if the majority of weight votes for it in tortoise. The only way this could happen is either a "51% attack" or else after a contentious hard fork. Such a node is effectively on a minority fork that won't have any valid blocks for a while.

Right. For a syntactically-invalid block, the node will ignore the Tortoise votes and drop the block (unless it needs to keep it just to provide a proof that the block creator's ID is invalid due to misbehavior).

> There should _never_ be two receipts for the same txid.

Also agree, but just to make sure we're on the same page --- a receipt is something internal in our implementation -- it's not extra data that "lives" on the blockchain, right?

**Comment by lrettig** [comment]
> a receipt is something internal in our implementation -- it's not extra data that "lives" on the blockchain, right?

Right, a receipt is not stored on the mesh, if that's what you mean. In the future it may be used by a light client protocol, or by a cross-chain/cross-shard protocol, but for now the main use case is for smapp, or other apps, to know whether a tx succeeded, how much gas it consumed, etc.

**Comment by dshulyak** [comment]
i am not sure what you mean when you are talking about tx syntactic validation. transaction validation **always** depends on the previous state. tx can be validated only after template was created and account is spawned with that template. so tx that included in layer N can be validated only when all txs from layer N - 1 were applied.

practically it means that if we want to verify that block is syntactically invalid we can do it only when previous layer was applied. this introduces significant changes to the tortoise, sync and the rest of the codebase.

just to summarize - "syntactic" validation for transactions is now practically impossible

**Comment by dshulyak** [comment]
changes for it to work during hdist

now: 
- certificate received for layer N 
- vote for the block

after:
- certificate received for layer N
- wait for layer N - 1 to be applied
- validate that transactions in the block are "synctactically" valid
- vote for the block

it is even worth for the sync because we will have to rescore goodness of the ballots **after** tortoise made a decision

now:
- download ballots
- tortoise outputs blocks
- apply blocks

after:
- download ballots
- tortoise outputs blocks
- validate block
- either: 
  - change tortoise opinion and rescore goodness
  - or apply blocks

**Comment by lrettig** [comment]
We briefly discussed this today. Summarizing my understanding here. @dshulyak @noamnelke please correct anything I've gotten wrong or missed here:

- we _can_ do some trivial syntactic (i.e., non-contextual) validation of transactions, even outside of the optimistic mode: effectively, just make sure a tx isn't "malformed" (wrong version byte, or too short to contain a principal address). there's nothing else to be checked as part of syntactic validation.
- we agree that a block containing one or more such "malformed" txs is itself "malformed", i.e., invalid, and per the above comments, this can only happen as the result of a 51% attack or another severe violation of the security assumptions and necessarily would result in a network fork of some kind
- the node does _not_ need to "wait for layer N-1 to be applied" to perform such basic syntactic validation
- it's certainly possible for "invalid" / "ineffective" transactions to appear in a block, including those with a principal that's nonexistent/cannot pay gas, and duplicate txs (consumed/applied in a previous block), due to optimistic filtering failing or other security assumption violations; in these cases the txs in question are ignored, don't pay gas, and the block containing them is _not_ considered invalid (I *think* all possible cases are covered in [this spreadsheet](https://docs.google.com/spreadsheets/d/1PNyKl4DuA17Ku85Ajl5A1lwkE9V7EXaYHJpkRxwyQ_U/edit))
- in future we may want to discuss modifying Hare protocol to vote against proposals containing txs that appear invalid, and/or sanctions against miners that include them, but this can be added later, after genesis, and doesn't require a hard fork

**Comment by dshulyak** [comment]
@lrettig there is an inconsistency in your document with what was described in this thread wrt max_gas and max_spend

> principal may have not funds to cover max_gas
This is a VALID transaction. So the principal pays as much gas as it can, and the tx fails if it runs out of gas (it's also reasonable to fail without executing, but I don't see a good reason to do so at the moment).

there is also a funny nuance about it. if self-spawn transaction fails in a such way (and zero nonce is consumed) after reorg then all other transaction from the same principals won't be executable as well.

also please add a condition for transactions that - `out of block gas`. i will assume that it is the same as with `fails intrinsic gas check`

**Comment by lrettig** [comment]
> there is an inconsistency in your document with what was described in this thread wrt max_gas and max_spend
> 
> > principal may have not funds to cover max_gas
> > This is a VALID transaction. So the principal pays as much gas as it can, and the tx fails if it runs out of gas (it's also reasonable to fail without executing, but I don't see a good reason to do so at the moment).

I assume you're referring to these in the spreadsheet:

> 4.2.2 |   |   | fails intrinsic gas check |   |   | principal account doesn't have enough balance to cover verify() + tx base fee + **max gas** | none
> 4.2.3 |   |   | not enough funds |   |   | principal account doesn't have enough balance to cover **max spend** | none

and the question is whether any gas should be paid in these scenarios? My opinion is that, in the first case (4.2.2, also 6.2.2, 7.2.2), no gas needs to be charged at all since this is a trivial check. In the second case (4.2.3, etc.), intrinsic gas can be charged (to pay for verify) but no portion of the "spend" amount should be charged.

@noamnelke do you agree?

> there is also a funny nuance about it. if self-spawn transaction fails in a such way (and zero nonce is consumed) after reorg then all other transaction from the same principals won't be executable as well.

I don't think there's anything we can do about this. If (e.g., after reorg) a principal account doesn't contain enough funds to cover the self-spawn, then of course the spawn can't happen and all subsequent txs would become invalid. In this case, a new spawn tx would have to be sent later after funds arrive (possibly with a higher nonce?)

> also please add a condition for transactions that - `out of block gas`

You mean like a block that exceeds the block gas limit? Wouldn't such a block be invalid in the first place?

**Comment by dshulyak** [comment]
> I don't think there's anything we can do about this. If (e.g., after reorg) a principal account doesn't contain enough funds to cover the self-spawn, then of course the spawn can't happen and all subsequent txs would become invalid. In this case, a new spawn tx would have to be sent later after funds arrive (possibly with a higher nonce?)

self-spawn can't set nonce explicitly, it is always zero by design

> You mean like a block that exceeds the block gas limit? Wouldn't such a block be invalid in the first place?

any information on gas is available only after parsing transactions. it is not possible to do for parsed transactions as discussed above

**Comment by dshulyak** [comment]
> and the question is whether any gas should be paid in these scenarios? My opinion is that, in the first case (4.2.2, also 6.2.2, 7.2.2), no gas needs to be charged at all since this is a trivial check. In the second case (4.2.3, etc.), intrinsic gas can be charged (to pay for verify) but no portion of the "spend" amount should be charged.

checking max_gas and max_spend is almost the same in terms of cost 

> intrinsic gas can be charged (to pay for verify) but no portion of the "spend" amount should be charged.

this part didn't came up before. in the document there is none. i will keep implementation consistent with a document

> tx is a duplicate of tx that's already included in a previous block

currently there is no special case for such type, it is assumed that such tx will be ineffective because of nonce validation

**Comment by lrettig** [comment]
> > I don't think there's anything we can do about this. If (e.g., after reorg) a principal account doesn't contain enough funds to cover the self-spawn, then of course the spawn can't happen and all subsequent txs would become invalid. In this case, a new spawn tx would have to be sent later after funds arrive (possibly with a higher nonce?)
> 
> self-spawn can't set nonce explicitly, it is always zero by design

Okay. This is a strange corner case and I can't think of a quick, easy workaround. In any case it should be rare. I'll open a separate thread/issue to investigate it.

> > You mean like a block that exceeds the block gas limit? Wouldn't such a block be invalid in the first place?
> 
> any information on gas is available only after parsing transactions. it is not possible to do for parsed transactions as discussed above

Couldn't we just use max_gas? This does not depend on execution context. In other words, a block with transactions with `max_gas` that sums to a number larger than the block gas limit is syntactically invalid. (This is not perfect since in theory a tx can specify a larger max_gas than it needs and occupy more block space than it should, but it won't be an issue at genesis and we can keep thinking about this problem.)

> checking max_gas and max_spend is almost the same in terms of cost

No, there is a difference. Checking max_gas is part of an "intrinsic gas check" that happens _before_ `verify()` is run. It's nearly costless. By contrast, checking max_spend only happens _after_ `verify()` is run. There's no point in checking max_spend before verifying the tx signature, and also, in future `verify()` may use a dynamic amount of gas. And the `verify()` operation is significantly more costly than the intrinsic gas check.

> > intrinsic gas can be charged (to pay for verify) but no portion of the "spend" amount should be charged.
> 
> this part didn't came up before. in the document there is none. i will keep implementation consistent with a document

Let's get @noamnelke to weigh in. It's pretty important that gas be paid if `verify()` is run or else there's a DoS vector. (I updated the spreadsheet to reflect this.)

> > tx is a duplicate of tx that's already included in a previous block
> 
> currently there is no special case for such type, it is assumed that such tx will be ineffective because of nonce validation

This is also an issue. Nonce validation does _not_ cover the case of a tx with a duplicate (i.e., previously consumed) txid. But we have a requirement that each txid be unique for all time - if a tx with the same txid appears later, it _must_ be ineffective (even if a template's nonce validation passes it).

**Comment by dshulyak** [comment]
> Couldn't we just use max_gas? This does not depend on execution context. In other words, a block with transactions with max_gas that sums to a number larger than the block gas limit is syntactically invalid. (This is not perfect since in theory a tx can specify a larger max_gas than it needs and occupy more block space than it should, but it won't be an issue at genesis and we can keep thinking about this problem.)

max_gas is available only after transaction is parsed, so we couldn't do what is suggested

> No, there is a difference. Checking max_gas is part of an "intrinsic gas check" that happens before verify() is run. It's nearly costless. By contrast, checking max_spend only happens after verify() is run. There's no point in checking max_spend before verifying the tx signature, and also, in future verify() may use a dynamic amount of gas. And the verify() operation is significantly more costly than the intrinsic gas check.

max_spend is parsed before calling verify, as it is more expensive than verify. technically max_spend can be even parsed during parse_payload call

> It's pretty important that gas be paid if verify() is run or else there's a DoS vector. (I updated the spreadsheet to reflect this.)

clearly gas can't be consumed if verify() doesn't pass, so not even sure what it changes from ddos perspective

**Comment by noamnelke** [comment]
Didn't read the entire thread, only last two messages, but I think I can help.

@lrettig is right that `MaxSpend()` is only executed after `Verify()`. It should be much cheaper, but it can only "block" future transactions, the decision to include this transaction in the mempool / proposal / block only depends on `MaxGas()` and `Verify()`. Once it's admitted, we want to check `MaxSpend()` so that we know how much balance is available for gas in following transactions. This is a tiny optimization that tries to minimize the work done before `Verify()`, but I disagree with @lrettig that it's critical--if doing calculating `MaxSpend()` as part of `ParsePayload()` saves us considerable complexity or makes execution more efficient, we should definitely consider it.

When we calculate the fixed gas cost of each method, each transaction will have a "base cost" that will be included in the method's gas cost. This is supposed to cover the cost of verification, as well as the cost of updating the nonce. There's no point in defining this cost separately. Let's look at what can happen:
- **Transaction doesn't pass verification** → No gas is charged
- **Transaction passes verification and has enough balance to cover the fixed gas cost** → The fixed gas cost is charged, covering verification and nonce update
- **Transaction passes verification, but only has balance to cover part of the fixed gas cost (even one smidge short)** → We don't attempt execution at all and use the entire balance as fee. Cost of verification and nonce update could be covered or even exceeded by this amount, or not, we charge it all anyway as "punishment" to incentivize users to avoid this situation as much as they can help it.
- I can't think of another scenario.

Duplicate transactions are covered by the nonce check: since the nonce doesn't depend on state, we'll detect duplicates that way. The one exception to this is a self-spawn transaction that failed due to insufficient balance to cover gas. I've given this some thought and we probably need to add a nonce to self spawn transactions after all, just to cover this case. I originally wrote that self-spawns don't need a nonce, when I was still planning to not consume transactions that can't cover gas.

`<rant>`
Consuming transactions still feels super-wrong to me. As I see it, smeshers can always include "garbage" in proposals / blocks because they can create a transaction with a made up principle or one that has zero balance. So we can't eliminate that ability, but the result of consuming transactions whose principle can't cover gas is that we can now also incur a processing cost for these transactions (since we have to run `Verify()` and the fee we're able to charge might not be able cover this cost). So as I see it, this just makes the problem worse for everyone (smeshers, validators and users - everyone is worse off).

I couldn't convince Tal of this, and I don't want to invest more time trying.
`</rant>`

**Comment by lrettig** [comment]
Summarizing some points from our call this morning:
- (contrary to what I said above) a block with txs that exceed the block gas limit is _not_ invalid. rather, any tx that exceeds the limit (i.e., starting with the first tx that exceeds it, and all of the rest of the txs in the block) are invalid.
- with respect to `max_spend`, the only thing the _system_ (VM) guarantees is that a tx does not spend more than `max_spend`. we do still execute a tx that doesn't have enough balance to cover its `max_spend`; what happens is up to the template/method.
- regarding a tx that cannot cover `max_gas`, @noamnelke and I still need to discuss this further with @tal-m, we will do so ASAP
- regarding duplicate txids, I posted some notes [here](https://spacemesh.slack.com/archives/CG91F13C7/p1658937507834559) on Slack. tldr: it might be expensive to check that txids are strictly unique inside the node/DB, so there is a corner case where a template that does _not_ perform a nonce check could result in duplicate txids. let's discuss further.
- regarding self-spawn txs, I think @noamnelke is right and we probably do need to add a nonce to these (I described the problem further [here](https://community.spacemesh.io/t/corner-case-ineffective-self-spawn-tx/287))
